pkg/metrics/metrics_test.go

```diff
@@ -353,3 +353,141 @@ func TestWriteMetric(t *testing.T) {
 		})
 	}
 }
+
+func TestWriteMetricAsync(t *testing.T) {
+	t.Parallel()
+
+	// Record calls made to test server. Separate per test using a per-test
+	// unique id in URL.
+	var reqMap sync.Map
+
+	// Request body is intentionally leaked to allow for inspection in test cases.
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Add 10ms of latency to response to allow testing timeout.
+		time.Sleep(10 * time.Millisecond)
+		saveReq := r.Clone(context.Background())
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			t.Fatalf("error copying body: %s", err.Error())
+		}
+		saveReq.Body = io.NopCloser(bytes.NewBuffer(body))
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+		_, prevExist := reqMap.Swap(r.RequestURI, saveReq)
+		if prevExist {
+			t.Fatalf("multiple requests to same url: %s", r.RequestURI)
+		}
+		if !strings.HasSuffix(r.RequestURI, "/sendMetrics") {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintln(w, http.StatusText(http.StatusNotFound))
+			return
+		}
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "ok")
+	}))
+
+	t.Cleanup(func() {
+		ts.Close()
+	})
+
+	cases := []struct {
+		name          string
+		metric        string
+		count         int64
+		client        *client
+		timeoutMillis int
+		wantRequest   *SendMetricRequest
+		wantErr       string
+	}{
+		{
+			name:   "metric_success",
+			metric: "foo",
+			count:  1,
+			client: defaultClient(),
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:          "metric_success_timeout_set",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 4000,
+			wantRequest: &SendMetricRequest{
+				AppID:      testAppID,
+				AppVersion: testVersion,
+				Metrics:    map[string]int64{"foo": 1},
+				InstallID:  testInstallID,
+			},
+		},
+		{
+			name:   "metric_opt_out_noop",
+			metric: "foo",
+			count:  1,
+			client: func() *client {
+				c := defaultClient()
+				c.OptOut = true
+				return c
+			}(),
+			wantRequest: nil,
+		},
+		{
+			name:          "metric_failure_timeout",
+			metric:        "foo",
+			count:         1,
+			client:        defaultClient(),
+			timeoutMillis: 9,
+			wantRequest:   nil,
+			wantErr:       "context deadline exceeded",
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx := context.Background()
+			if tc.timeoutMillis > 0 {
+				var done func()
+				ctx, done = context.WithTimeout(ctx, time.Duration(tc.timeoutMillis)*time.Millisecond)
+				defer done()
+			}
```

Comment:
It might be better to just accept a `context.Context` instead in the cases struct

Reply:
Good point